test/log.ts

```diff
@@ -413,161 +435,157 @@ describe('Log', () => {
+          undefined
+        )
+      );
     });
 
     it('should call gax method', async () => {
-      log.logging.loggingService.writeLogEntries = (
-        reqOpts: {},
-        gaxOpts: {}
-      ) => {
-        assert.deepStrictEqual(reqOpts, {
-          logName: log.formattedName_,
-          entries: [ENTRY],
-          resource: FAKE_RESOURCE,
-        });
-
-        assert.strictEqual(gaxOpts, undefined);
-      };
-
       await log.write(ENTRY, OPTIONS);
+      assert(
+        writeLogStub.calledOnceWithExactly(
+          {
+            logName: log.formattedName_,
+            entries: [ENTRY],
+            resource: FAKE_RESOURCE,
+          },
+          undefined
+        )
+      );
     });
 
     it('should arrify & decorate the entries', async () => {
       const decoratedEntries = [] as Entry[];
-
-      log.decorateEntries_ = (entries: Entry[]) => {
-        assert.strictEqual(entries[0], ENTRY);
-        return decoratedEntries;
-      };
-
-      // tslint:disable-next-line no-any
-      log.logging.loggingService.writeLogEntries = (
-        // tslint:disable-next-line no-any
-        reqOpts: any,
-        gaxOpts: {}
-      ) => {
-        assert.strictEqual(reqOpts.entries, decoratedEntries);
-      };
+      decorateEntriesStub.returns(decoratedEntries);
+      const arrifiedEntries = [ENTRY] as Entry[];
 
       await log.write(ENTRY, OPTIONS);
+      assert(decorateEntriesStub.calledOnceWithExactly(arrifiedEntries));
+      assert(
+        writeLogStub.calledOnceWith(
+          sinon.match({
+            entries: arrifiedEntries,
+          })
+        )
+      );
     });
 
     it('should not require options', async () => {
-      log.logging.loggingService.writeLogEntries = (
-        reqOpts: {},
-        gaxOpts: {}
-      ) => {};
-
       await log.write(ENTRY);
+      assert(writeLogStub.calledOnceWithExactly(sinon.match.object, undefined));
     });
 
     it('should pass through additional options', async () => {
-      log.logging.loggingService.writeLogEntries = (
-        reqOpts: WriteOptions,
-        gaxOpts: {}
-      ) => {
-        assert.strictEqual(reqOpts.dryRun, true);
-        assert.strictEqual(reqOpts.partialSuccess, false);
-      };
-
       await log.write(ENTRY, {dryRun: true, partialSuccess: false});
+      assert(
+        writeLogStub.calledOnceWith(
+          sinon.match({
+            dryRun: true,
+            partialSuccess: false,
+          })
+        )
+      );
     });
 
-    it('should not truncate entries by default', async () => {
-      const logger = createLogger();
-      const entry = new Entry({}, 'hello world'.padEnd(300000, '.'));
-      // tslint:disable-next-line no-any
-      logger.logging.loggingService.writeLogEntries = (
-        // tslint:disable-next-line no-any
-        reqOpts: any,
-        _gaxOpts: {}
-      ) => {
-        assert.strictEqual(reqOpts.entries[0].textPayload.length, 300000);
-      };
+    describe('truncateEntries', () => {
```

Comment:
While you're in here, could you move this out to a root-level describe? Since it's its own method, it should be tested separately. I would move it from being `private truncateEntries` to just `truncateEntries_` to match our conventions.

No problem if you prefer that work stay out of the scope of this PR.

Reply:
I had thought about it! Depends on our testing philosophy: do we unit test "private" methods (all the things!), or do we test public contracts and allow later refactoring of the implementation (functional)?
I'll look into what other parts of the code rely on truncation, and if it's just used here I think we should leave it as a nested suite.

Reply:
Our functional tests are just the system tests, otherwise, our tests essentially go function-by-function (every one!), line-by-line. This nested suite is the first deviance I've seen across our libraries (I don't think I reviewed the PR it was introduced in), so it's possible it's slipped in. But, I would prefer we stick to the ease of a straight mapping between function -> test.